Fix parallel variant ordering clash

thecrypticace opened the pull request:
Our ordering for parallel variants could result in some non-deterministic behavior depending on which ones you used and how many utilities were registered before that utility that was using the parallel variant.

This was a side effect of the single bigint sorting scheme we used. We've replaced it with Objects that have a better breakdown of what each section represents. This also means that the bigint numbers we're working with are now MUCH smaller. And, generally, only have to use large numbers for the bitfield used for variants.

This should also result in a minor speedup for the implementation of `getClassOrder`

thecrypticace committed:
Remove remnants of the user layer

It hasn’t been used in a while

thecrypticace committed:
Rewrite sort offset generation

thecrypticace committed:
wip

thecrypticace committed:
wip

wip

thecrypticace committed:
Handle parasite utilities

thecrypticace committed:
wip

thecrypticace committed:
wip

thecrypticace committed:
Make parallel variants sorting more resillient

It’s not perfect but it’s close

thecrypticace committed:
fix

thecrypticace committed:
remove todo

it adds a new bit so it can’t

thecrypticace committed:
Simplify getClassOrder usage

thecrypticace committed:
Simplify

oops

oops

thecrypticace committed:
Add parasite utility for `dark`

dark mode class name

thecrypticace committed:
Cleanup

thecrypticace committed:
Cleanup

thecrypticace committed:
Simplify

RobinMalfait committed:
format files

RobinMalfait commented on tests/arbitrary-variants.test.js:
```diff
@@ -107,7 +107,7 @@ test('variants without & or an at-rule are ignored', () => {
 test('arbitrary variants are sorted after other variants', () => {
   let config = {
-    content: [{ raw: html`<div class="[&>*]:underline underline lg:underline"></div>` }],
+    content: [{ raw: html`<div class="underline lg:underline [&>*]:underline"></div>` }],
```
I had an error in my terminal and fixed it and now GitHub Actions is complaining and I'm pretty sure the new code is the correct code. 🤔

thecrypticace committed:
Fix prettier plugin to use git build of Tailwind CSS

Symlink and build instead of adding a recursive dev dependency

It breaks node < 16

thecrypticace committed:
Fix prettier error

thecrypticace committed:
wip

thecrypticace committed:
fix test

thecrypticace committed:
Update changelog

thecrypticace committed:
Merge branch 'master' into fix/parallel-variant-ordering-clash

# Conflicts:
#	CHANGELOG.md

AndyOGo commented:
Changing the order of generated classes affects precedence.

Why is this not a breaking change?

RobinMalfait commented:
@AndyOGo are you running into actual issues? The ordering changes we did shouldn't result in any breaking change unless you were already relying on behaviour that was non-deterministic before and could break the moment you introduced a new content file. A more detailed comment can be found here: https://github.com/tailwindlabs/tailwindcss/issues/10603#issuecomment-1433743390

If you are running into issues, then could you open an issue with a minimal reproduction repo attached so that we can take a look?

AndyOGo commented:
Thanks for your quick reply and sharing that comment @RobinMalfait 
I understand your point of view.

I look at it from a CSS spec and semantic versioning viewpoint.
Changing the order of classes, changes the precedence of applying those style rules.
For that reason it's a breaking change.
